packages/flutter/lib/src/services/text_input.dart

```diff
@@ -1130,7 +1288,7 @@ RawFloatingCursorPoint _toTextPoint(FloatingCursorDragState state, Map<String, d
 class TextInput {
   TextInput._() {
     _channel = SystemChannels.textInput;
-    _channel.setMethodCallHandler(_handleTextInputInvocation);
+    _channel.setMethodCallHandler(_TextInputChannelConnection._handleTextInputInvocation);
```

Comment:
any chance we can set the handler when we attach to a new connection and unset when we detach from this connection? It sounds wrong that when we switch connection the old handler is still listening to the text input channel

Reply:
I recall we had it like that in the beginning, but there was a problem with how `_DefaultBinaryMessenger` deal with unhandled messages: https://github.com/flutter/flutter/pull/69146#discussion_r523240713

Reply:
Ah right thanks for pointing that out. It feels like we should let `TextInputConnection` own the text input channel (at least pretend to own it since it's a singleton), and decide what to do when it disconnects? Now that `TextInputConnection` handles the communication between the client and the keyboard,  should `TextInput` just be responsible for creating/switching connections (attach, detach)? It's OK to leave `finishAutofillContext` in `TextInput` for now since we can still access the method channel via `SystemChannels.textInput`. WDYT?

Reply:
We are still missing a good place for static `TextInputConnection` init and cleanup. The upcoming `TextInputSource` could help with that because it could have such methods that init and cleanup things that apply to all specific type of input connection instances.

The default `TextInputSource` implementation could act as the "owner" of the input channel and pass it over to input connections. It could also set a stub method call handler that discards all incoming messages when not in use.

> It's OK to leave finishAutofillContext in TextInput for now since we can still access the method channel via SystemChannels.textInput.

Accessing `SystemChannels.textInput` directly would unfortunately break autofill tests that rely on `TextInput.setChannel()`. We could actually solve this problem too with `TextInputSource`, though. Since `finishAutofillContext()` is not tied to any specific input connection, we could redirect it to the current input source.

Reply:
I added an internal `_TextInputSourrce` to address these issues. I'll follow up with the rest in #69146.